File: mobile-sdk/ios/Sources/CdeSDK/CdeSDK.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// A document ready to render, with the branch already taken.
public enum OpenedDocument: Sendable {
    /// The URL is nil only when opened with `prefetch: false` and not cached.
    case pdf(ViewerSource.PdfSource, URL?)
    case drawing(ViewerSource.DrawingSource)
    case image(ViewerSource.ImageSource)
    case unsupported(ViewerSource.UnsupportedSource)
}

/// The entry point to the CDE SDK.
///
/// One instance per signed-in user, held for the life of the app:
///
/// ```swift
/// let cde = CdeSDK(configuration: .init(baseURL: URL(string: "https://cde.example.com")!))
/// try await cde.signIn(username: "…", password: "…")
/// let documents = try await cde.documents(projectId: 1)
/// ```
///
/// Everything that touches the network answers from the offline cache when
/// there is none, so a caller does not branch on connectivity — the
/// alternative is every screen in the host app reimplementing that choice,
/// differently.
public actor CdeSDK {

    private let api: CdeAPI
    private let tokens: TokenStore
    private let store: OfflineStore
    private let sync: SyncEngine

    public init(configuration: CdeConfiguration, storageDirectory: URL? = nil) {
        let tokens = TokenStore()
        let store = OfflineStore(directory: storageDirectory)

        self.tokens = tokens
        self.store = store
        self.api = CdeAPI(
            configuration: configuration,
            tokenProvider: { tokens.token },
            onUnauthenticated: { tokens.clear() })
        self.sync = SyncEngine(api: api, store: store)
    }

    public var isSignedIn: Bool { tokens.token != nil }
    public var signedInAs: String? { tokens.username }

    /// Queue depth and progress, for a status indicator in the host app.
    public func observeSyncState(_ handler: @escaping @Sendable (SyncState) -> Void) async {
        await sync.observeState(handler)
    }

    // MARK: - Session

    @discardableResult
    public func signIn(username: String, password: String) async throws -> AuthResult {
        let result = try await api.login(username: username, password: password)
        tokens.store(token: result.token, username: result.username, role: result.role)
        return result
    }

    /// Ends the session.
    ///
    /// Cached documents go; the outbound queue does not, unless
    /// `discardPendingChanges` is set. Markup someone drew on site exists
    /// nowhere else, and signing out — often just to switch account — is not a
    /// statement that they wanted to throw it away.
    public func signOut(discardPendingChanges: Bool = false) {
        tokens.clear()
        store.clearCachedContent()
        if discardPendingChanges { store.replace(with: []) }
    }

    // MARK: - Browsing

    public func projects() async throws -> [Project] {
        try await api.projects()
    }

    /// Documents in a project, falling back to the last list seen offline.
    public func documents(projectId: Int64) async throws -> [CdeDocument] {
        do {
            let documents = try await api.documents(projectId: projectId)
            store.storeDocuments(documents, projectId: projectId)
            return documents
        } catch CdeError.offline {
            return store.cachedDocuments(projectId: projectId)
        }
    }

    // MARK: - Opening a document

    /// How to render a document, and — for a PDF — the file to render from,
    /// downloaded if it is not already cached.
    ///
    /// - Parameter prefetch: when false the PDF is not fetched, so a document
    ///   list can ask cheaply what something is.
    public func open(documentId: Int64, prefetch: Bool = true) async throws -> OpenedDocument {
        let source = try await api.viewerSource(documentId: documentId)

        switch source {
        case .pdf(let pdf):
            // The URL carries ?v=, so it is also the cache key: a new version
            // cannot be served from an old file.
            if let cached = store.cachedFile(for: pdf.path) {
                return .pdf(pdf, cached)
            }
            guard prefetch else { return .pdf(pdf, nil) }
            let data = try await api.pdfData(path: pdf.path)
            return .pdf(pdf, try store.store(data, for: pdf.path))

        case .drawing(let drawing): return .drawing(drawing)
        case .image(let image): return .image(image)
        case .unsupported(let unsupported): return .unsupported(unsupported)
        }
    }

    // MARK: - Annotations

    public func annotations(documentId: Int64) async throws -> [Annotation] {
        try await sync.annotations(documentId: documentId)
    }

    /// Records markup.
    ///
    /// Queued locally first and sent when possible, so the call succeeds with
    /// no signal and the drawing is never lost to a failed request. The
    /// returned annotation carries a negative id until the server assigns a
    /// real one — callers should key on `shapeData.id`, which is stable from
    /// the moment it is drawn.
    @discardableResult
    public func addAnnotation(
        documentId: Int64,
        shape: ShapeData,
        type: AnnotationType = .markup,
        comment: String? = nil
    ) async throws -> Annotation {
        let request = AnnotationRequest(
            documentId: documentId,
            type: type,
            shapeData: try MarkupCodec.encode(shape),
            comment: comment,
            pageNumber: shape.pageNumber)

        store.enqueue(.create(localId: shape.id, request: request, attempts: 0))
        await sync.sync()

        return Annotation(
            id: -1, documentId: documentId, type: type,
            shapeData: request.shapeData, comment: comment,
            pageNumber: shape.pageNumber)
    }

    public func resolveAnnotation(id: Int64) async {
        store.enqueue(.resolve(localId: "resolve-\(id)", annotationId: id, attempts: 0))
        await sync.sync()
    }

    public func deleteAnnotation(id: Int64) async {
        store.enqueue(.delete(localId: "delete-\(id)", annotationId: id, attempts: 0))
        await sync.sync()
    }

    /// Sends anything queued. Call on foreground and when connectivity
    /// returns; it is safe to call more often than necessary.
    ///
    /// - Returns: changes the server refused outright, which will not be
    ///   retried.
    @discardableResult
    public func synchronise() async -> [RejectedChange] {
        await sync.sync()
    }

    // MARK: - Storage

    public func cacheSizeBytes() -> Int64 { store.cacheSizeBytes() }

    public func trimCache(to limit: Int64) { store.trim(to: limit) }
}

File: mobile-sdk/android/cde-sdk/src/main/kotlin/com/cde/sdk/CdeSdk.kt
package com.cde.sdk

import android.content.Context
import com.cde.sdk.auth.TokenStore
import com.cde.sdk.model.*
import com.cde.sdk.net.AuthResult
import com.cde.sdk.net.CdeApi
import com.cde.sdk.net.CdeConfiguration
import com.cde.sdk.net.CdeError
import com.cde.sdk.offline.OfflineStore
import com.cde.sdk.offline.PendingChange
import com.cde.sdk.offline.SyncEngine
import com.cde.sdk.render.PdfPageRenderer
import kotlinx.coroutines.flow.StateFlow
import java.io.File

/**
 * The entry point to the CDE SDK.
 *
 * One instance per signed-in user, held for the life of the app:
 *
 * ```kotlin
 * val cde = CdeSdk(context, CdeConfiguration(baseUrl = "https://cde.example.com"))
 * cde.signIn("username", "password")
 * val documents = cde.documents(projectId = 1).content
 * ```
 *
 * Everything that touches the network suspends and answers from the offline
 * cache when there is none, so a caller does not branch on connectivity —
 * the alternative is every screen in the host app reimplementing that choice,
 * differently.
 */
class CdeSdk(
    context: Context,
    configuration: CdeConfiguration,
) {
    private val applicationContext = context.applicationContext
    private val tokens = TokenStore(applicationContext)
    private val store = OfflineStore(applicationContext)

    private val api = CdeApi(
        configuration = configuration,
        tokenProvider = { tokens.token },
        onUnauthenticated = { tokens.clear() },
    )

    private val sync = SyncEngine(api, store)

    /** Queue depth and progress, for a status indicator in the host app. */
    val syncState: StateFlow<com.cde.sdk.offline.SyncState> get() = sync.state

    val isSignedIn: Boolean get() = tokens.token != null
    val signedInAs: String? get() = tokens.username

    // ── Session ──────────────────────────────────────────────────

    suspend fun signIn(username: String, password: String): AuthResult {
        val result = api.login(username, password)
        tokens.store(result.token, result.username, result.role)
        return result
    }

    /**
     * Ends the session.
     *
     * Cached documents go; the outbound queue does not, unless
     * [discardPendingChanges] is set. Markup someone drew on site exists
     * nowhere else, and signing out — often just to switch account — is not
     * a statement that they wanted to throw it away.
     */
    fun signOut(discardPendingChanges: Boolean = false) {
        tokens.clear()
        store.clearCachedContent()
        if (discardPendingChanges) store.replace(emptyList())
    }

    // ── Browsing ─────────────────────────────────────────────────

    suspend fun projects(): List<Project> = api.projects()

    /**
     * One page of a project's documents, falling back to what was last seen
     * when there is no connection.
     *
     * <p>The cache holds the first page only, and the page returned offline
     * describes itself as exactly that — the totals are the cache's own, not a
     * remembered server count, so a caller is never told there are 137
     * documents when it can produce 50.
     */
    suspend fun documents(projectId: Long, page: Int = 0, size: Int = 50): DocumentPage = try {
        api.documents(projectId, page, size)
            .also { if (it.number == 0) store.storeDocuments(projectId, it.content) }
    } catch (e: CdeError.Offline) {
        val cached = if (page == 0) store.cachedDocuments(projectId) else emptyList()
        DocumentPage(
            content = cached, number = page, size = size,
            totalElements = cached.size.toLong(), totalPages = 1,
            first = page == 0, last = true)
    }

    // ── Opening a document ───────────────────────────────────────

    /**
     * How to render a document, and — for a PDF — the file to render from,
     * downloaded if it is not already cached.
     *
     * @param prefetch when true the PDF is fetched even if the caller only
     *        meant to inspect the type. Left false so a document list can ask
     *        cheaply what something is.
     */
    suspend fun open(documentId: Long, prefetch: Boolean = true): OpenedDocument {
        val source = api.viewerSource(documentId)
        return when (source) {
            is ViewerSource.Pdf -> {
                // The URL carries ?v=, so it is also the cache key: a new
                // version cannot be served from an old file.
                val cached = store.cachedFile(source.pdfPath)
                val file = when {
                    cached != null -> cached
                    prefetch -> store.storeFile(source.pdfPath, api.pdfBytes(source.pdfPath))
                    else -> null
                }
                OpenedDocument.Pdf(source, file)
            }

            is ViewerSource.Drawing -> OpenedDocument.Drawing(source)
            is ViewerSource.Image -> OpenedDocument.Image(source)
            is ViewerSource.Unsupported -> OpenedDocument.Unsupported(source)
        }
    }

    /** Opens a cached PDF for rendering. Caller closes it. */
    fun renderer(file: File): PdfPageRenderer = PdfPageRenderer.open(file)

    // ── Annotations ──────────────────────────────────────────────

    suspend fun annotations(documentId: Long): List<CdeAnnotation> = sync.annotations(documentId)

    /**
     * Records markup.
     *
     * Queued locally first and sent when possible, so the call succeeds with
     * no signal and the drawing is never lost to a failed request. The
     * returned annotation carries a negative id until the server assigns a
     * real one — callers should key on `shapeData.id`, which is stable from
     * the moment it is drawn.
     */
    suspend fun addAnnotation(
        documentId: Long,
        shape: ShapeData,
        type: AnnotationType = AnnotationType.MARKUP,
        comment: String? = null,
    ): CdeAnnotation {
        val request = AnnotationRequest(
            documentId = documentId,
            type = type,
            shapeData = MarkupCodec.encode(shape),
            comment = comment,
            pageNumber = shape.pageNumber,
        )
        store.enqueue(PendingChange.Create(localId = shape.id, request = request))
        sync.sync()
        return CdeAnnotation(
            id = -1, documentId = documentId, type = type,
            shapeData = request.shapeData, comment = comment,
            pageNumber = shape.pageNumber,
        )
    }

    suspend fun resolveAnnotation(annotationId: Long) {
        store.enqueue(PendingChange.Resolve(
            localId = "resolve-$annotationId", annotationId = annotationId))
        sync.sync()
    }

    suspend fun deleteAnnotation(annotationId: Long) {
        store.enqueue(PendingChange.Delete(
            localId = "delete-$annotationId", annotationId = annotationId))
        sync.sync()
    }

    /**
     * Sends anything queued. Call on app resume and when connectivity
     * returns; it is safe to call more often than necessary.
     *
     * @return changes the server refused outright, which will not be retried
     */
    suspend fun synchronise(): List<SyncEngine.RejectedChange> = sync.sync()

    // ── Storage ──────────────────────────────────────────────────

    fun cacheSizeBytes(): Long = store.cacheSizeBytes()

    fun trimCacheTo(limitBytes: Long) = store.trimTo(limitBytes)
}

/** A document ready to render, with the branch already taken. */
sealed interface OpenedDocument {
    val source: ViewerSource

    data class Pdf(
        override val source: ViewerSource.Pdf,
        /** Null only when opened with `prefetch = false` and not cached. */
        val file: File?,
    ) : OpenedDocument

    data class Drawing(override val source: ViewerSource.Drawing) : OpenedDocument
    data class Image(override val source: ViewerSource.Image) : OpenedDocument
    data class Unsupported(override val source: ViewerSource.Unsupported) : OpenedDocument
}
